ledger/participant-integration-api/src/main/scala/platform/index/IndexServiceImpl.scala

```diff
@@ -116,17 +116,30 @@ private[index] class IndexServiceImpl(
         dispatcher()
           .startingAt(
             from.getOrElse(Offset.beforeBegin),
-            RangeSource { (startExclusive, endInclusive) =>
-              filterSource(transactionFilter, verbose) {
-                (templateFilter, eventProjectionProperties) =>
-                  transactionsReader
-                    .getFlatTransactions(
-                      startExclusive,
-                      endInclusive,
-                      templateFilter,
-                      eventProjectionProperties,
-                    )
-              }
+            RangeSource {
+              @volatile var metadata: PackageMetadata = null
+              @volatile var source
+                  : Source[(Map[Party, Set[Identifier]], EventProjectionProperties), NotUsed] =
+                Source.empty
+
+              (startExclusive, endInclusive) =>
+                val currentMetadata = packageMetadataView.current()
+                if (metadata ne currentMetadata) {
```

Comment:
I am wondering if this logic can be encapsulated by some existing primitives or caches. Still looks ugly and quite technical.